### PR TITLE
Added merge tests for custom styles

### PR DESCRIPTION
**Pull request by GeoffCoxMSFT**
## Changes
- Added test for shallow merge of customStyleHooks in FluentProvider

## Issues
- Updates #27058

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                                             |          Size (minified/GZIP) |
| ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ButtonProviderAndTheme.fixture.js'>react-components: Button, FluentProvider & webLightTheme</abbr>              |  `67.061 kB`<br />`18.105 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `205.644 kB`<br />`57.056 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr>                            |  `37.891 kB`<br />`12.039 kB` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr>                                                         |    `6.324 kB`<br />`2.129 kB` |
| <samp>react-provider</samp> <br /> <abbr title='bundle-size/FluentProvider.fixture.js'>FluentProvider</abbr>                                                                  |   `19.958 kB`<br />`6.838 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/efb2274e37c0237f3a10a4630b2696a9880f6398'>efb2274e37c0237f3a10a4630b2696a9880f6398</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/Avatar-mount_base.html">1582</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/Avatar-mount.html">1594</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/Button-mount_base.html">1126</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/Button-mount.html">1126</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/Field-mount_base.html">2419</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/Field-mount.html">2438</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/FluentProvider-mount_base.html">1951</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/FluentProvider-mount.html">1926</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/FluentProviderWithTheme-mount_base.html">757</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/FluentProviderWithTheme-mount.html">768</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">696</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">690</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">764</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">750</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>InfoButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/InfoButton-mount_base.html">675</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/InfoButton-mount.html">674</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/MakeStyles-mount_base.html">2250</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/MakeStyles-mount.html">2238</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/Persona-mount_base.html">3695</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/Persona-mount.html">3586</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/SpinButton-mount_base.html">2905</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27086/perf-test-react-components/SpinButton-mount.html">2880</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluentuiv9"/>

**Review by khmakoto — APPROVED** [review]

File: packages/react-components/react-provider/src/components/FluentProvider/useFluentProvider.test.tsx [hunk 1]
```diff
@@ -81,4 +82,67 @@ describe('useFluentProvider_unstable', () => {
...
+      useCompoundButtonStyles_unstable: (state: unknown) => {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (state as any).testResult = 'useCompoundButtonStyles_unstable_B';
+      },
+      // eslint-disable-next-line @typescript-eslint/naming-convention
+      useSplitButtonStyles_unstable: (state: unknown) => {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (state as any).testResult = 'useSplitButtonStyles_unstable_B';
+      },
+    };
```
Comment (layershifter) [inline packages/react-components/react-provider/src/components/FluentProvider/useFluentProvider.test.tsx:111]: nit: You don't need `state` to be used there actually, you can just return a string:

```suggestion
    const customStylesA: FluentProviderCustomStyleHooks = {
      // eslint-disable-next-line @typescript-eslint/naming-convention
      useButtonStyles_unstable: () => {
        return 'useButtonStyles_unstable_A';
      },
      // eslint-disable-next-line @typescript-eslint/naming-convention
      useCompoundButtonStyles_unstable: () => {
        return 'useCompoundButtonStyles_unstable_A';
      },
    };

    const customStylesB: FluentProviderCustomStyleHooks = {
      // eslint-disable-next-line @typescript-eslint/naming-convention
      useCompoundButtonStyles_unstable: () => {
        return 'useCompoundButtonStyles_unstable_B';
      },
      // eslint-disable-next-line @typescript-eslint/naming-convention
      useSplitButtonStyles_unstable: () => {
        return 'useSplitButtonStyles_unstable_B';
      },
    };
```

File: packages/react-components/react-provider/src/components/FluentProvider/useFluentProvider.test.tsx [hunk 2]
```diff
@@ -81,4 +82,67 @@ describe('useFluentProvider_unstable', () => {
...
+      <FluentProvider customStyleHooks_unstable={customStylesA}>{children}</FluentProvider>
+    );
+
+    const { result } = renderHook(
+      // eslint-disable-next-line @typescript-eslint/naming-convention
+      () => useFluentProvider_unstable({ customStyleHooks_unstable: customStylesB }, React.createRef()),
+      {
+        wrapper: Wrapper,
+      },
+    );
```
Comment (layershifter) [inline packages/react-components/react-provider/src/components/FluentProvider/useFluentProvider.test.tsx:133]: Let's split this to two separate `it` statements as TBH I am lost in assertions below.

**Review by layershifter — APPROVED** [review]

**Pull request merged**